src/enzo/Grid_ApplyGalaxyParticleGravity.C

```diff
@@ -0,0 +1,117 @@
+#include "phys_constants.h"
+#include "ExternalBoundary.h"
+#include "Grid.h"
+#include "Hierarchy.h"
+#include "ActiveParticle.h"
+#include "phys_constants.h"
+#include "ActiveParticle_GalaxyParticle.h"
+
+#define NO_DEBUG
+
+int grid::ApplyGalaxyParticleGravity(ActiveParticleType** ThisParticle){
+
+  /* Return if this doesn't involve us */
+  if (MyProcessorNumber != ProcessorNumber) 
+    return SUCCESS;
+
+  /* Check whether the cube that circumscribes the accretion zone intersects with this grid */
+
+   FLOAT *ParticlePosition = (*ThisParticle)->ReturnPosition();
+   float rad = static_cast<ActiveParticleType_GalaxyParticle*>(*ThisParticle)->Radius;
+   
+   if ((GridLeftEdge[0] > ParticlePosition[0]+rad) || (GridRightEdge[0] < ParticlePosition[0]-rad) ||
+       (GridLeftEdge[1] > ParticlePosition[1]+rad) || (GridRightEdge[1] < ParticlePosition[1]-rad) ||
+       (GridLeftEdge[2] > ParticlePosition[2]+rad) || (GridRightEdge[2] < ParticlePosition[2]-rad))
+     return SUCCESS;
+
+  
+  int i, j, k, index;
+  float rad2dx, dist2;
+  float dx = float(this->CellWidth[0][0]);
+  float max_gmf = 0.;
+  FLOAT xx, yy, zz;
+
+  int SNColourNum, MetalNum, MBHColourNum, Galaxy1ColourNum, Galaxy2ColourNum,
+    MetalIaNum, MetalIINum;
+  int MetallicityField = FALSE;
+
+  if (this->IdentifyColourFields(SNColourNum, MetalNum, MetalIaNum,
+                                 MetalIINum, MBHColourNum, 
+                                 Galaxy1ColourNum, Galaxy2ColourNum) == FAIL) {
+    ENZO_FAIL("Error in grid->IdentifyColourFields.\n");
+  }
+  MetallicityField = (MetalNum > 0) ? TRUE : FALSE;
+
+  int DensNum, GENum, TENum, Vel1Num, Vel2Num, Vel3Num;
+  if (this->IdentifyPhysicalQuantities(DensNum, GENum, Vel1Num, Vel2Num,
+         Vel3Num, TENum) == FAIL) {
+     ENZO_FAIL("Error in IdentifyPhysicalQuantities.\n");
+   }
+
+  rad2dx = (rad + dx) * (rad + dx);
+  
+  printf("I would be applying gravity to this grid:\n");
+  printf("%f %f %f : %f %f %f\n", GravitatingMassFieldLeftEdge[0],
+    GravitatingMassFieldLeftEdge[1], GravitatingMassFieldLeftEdge[2],
+    GravitatingMassFieldLeftEdge[0] + GravitatingMassFieldDimension[0]* GravitatingMassFieldCellSize,
+    GravitatingMassFieldLeftEdge[1] + GravitatingMassFieldDimension[1]* GravitatingMassFieldCellSize,
+    GravitatingMassFieldLeftEdge[2] + GravitatingMassFieldDimension[2]* GravitatingMassFieldCellSize);
+//   printf("%d %d %d : %d %d %d\n", this->GridStartIndex[0], this->GridStartIndex[1],
+//     this->GridStartIndex[2], this->GridEndIndex[0], this->GridEndIndex[1],
+//     this->GridEndIndex[2]);
+//   printf("%f %f %f\n", this->CellLeftEdge[0][3] + 0.5*this->CellWidth[0][3],
+//     this->CellLeftEdge[1][3] + 0.5*this->CellWidth[1][3],
+//     this->CellLeftEdge[2][3] + 0.5*this->CellWidth[2][3]);
+  
+    for (k = 0; k < GravitatingMassFieldDimension[0]; k++) {
+      for (j = 0; j < GravitatingMassFieldDimension[1]; j++) {
+        index = GRIDINDEX_NOGHOST(0, j, k);
```

Comment:
Hard-coded 0 instead of GridStartIndex[0]. I'm not sure if this is always equivalent, but might be better to switch to the more general version.

Reply:
I think this has to be like this as it's outside the "i" loop

Reply:
Just to clarify - I mean replacing the 0 with "GridStartIndex[0]", not "i".  I suspect that in most cases, GridStartIndex[0] = 0 and this is equivalent anyway

Reply:
Ah sorry yes. I think you're correct GridStarIndex[0] would be more general

Reply:
We're talking about the 0 in GRIDINDEX_NOGHOST(o,j,k)?  I think this intended to be a loop over all values, so should start at 0 (just like the j and k loops above).  Note that GridStartIndex is the start of the active region and so generally not 0 (i.e. it is the first non-ghost zone).  So I think this code as originally written is correct...  (but ignore me if I'm talking nonsense!).

Reply:
You're right! I was conflating different use cases. This looks good to me!